### PR TITLE
Bug: accessed state outputs should not raise if also an argument or option

### PR DESCRIPTION
**Pull request by tbconroy**

### CODE REVIEW

File: lib/flow/state/output.rb [hunk 1]
```diff
@@ -38,7 +38,9 @@ def output(output, default: nil, &block)
 
         def ensure_validation_before(method)
           around_method method do |*arguments|
-            raise NotValidatedError unless validated?
+            method_name = method.to_s.delete("=").to_sym
+
+            raise NotValidatedError unless validated? || _options.include?(method_name) || _arguments.include?(method_name)
```
Comment (tbconroy) [inline lib/flow/state/output.rb:43]: Empty `_arguments` is an empty hash

File: lib/flow/state/output.rb [hunk 2]
```diff
@@ -38,7 +38,9 @@ def output(output, default: nil, &block)
 
         def ensure_validation_before(method)
           around_method method do |*arguments|
-            raise NotValidatedError unless validated?
+            method_name = method.to_s.delete("=").to_sym
```
Comment (tbconroy) [inline lib/flow/state/output.rb:41]: We don't want it to raise on the setter

**Comment by tbconroy** [comment]
@arettberg I notice that exact copies of`Flow::State::Output` and `Flow::State::Status` exist in Spicerack, there is a convenient class `SpiceRack::OutputObject` that includes these models, https://github.com/Freshly/spicerack/blob/develop/lib/spicerack/output_object.rb.

Should the real fix here be:
1) update `Spicerack::Objects::Output` to allow outputs that are also arguments or options to not raise https://github.com/Freshly/spicerack/blob/develop/lib/spicerack/objects/output.rb
2) update `Flow::StateBase` to inherit from `SpiceRack::OutputObject` instead of `Spicerack::InputModel`

**Comment by tbconroy** [comment]
@arettberg Another thought I had:
Should we allow the following new option for `Spicerack::Objects::Arguments` and `Spicerack::Objects::Options` for method definitions:
https://github.com/Freshly/spicerack/blob/develop/lib/spicerack/objects/arguments.rb
https://github.com/Freshly/spicerack/blob/develop/lib/spicerack/objects/options.rb

I'm thinking allow for an `output: true` option to be passed in the `option` and `argument` class method that would add the method name to `_options` under the hood.
```
argument :bar, output: true
option :foo, output: true
```

**Comment by arettberg** [comment]
@tbconroy we can't do that. Those modules are used by objects that aren't output objects, so that would create some pretty hairy interdependencies

File: lib/flow/state/output.rb (hunk 2 above)
Comment (arettberg) [reply lib/flow/state/output.rb:41]: why not? also, if this is really what we want, why to just remove  line 36?

File: lib/flow/state/output.rb (hunk 1 above)
Comment (arettberg) [reply lib/flow/state/output.rb:43]: ?

Comment (tbconroy) [reply lib/flow/state/output.rb:43]: Reason for why this line isn't something like `(_options + _arguments).include?(method_name)`

File: lib/flow/state/output.rb (hunk 2 above)
Comment (tbconroy) [reply lib/flow/state/output.rb:41]: > We don't want it to raise on the setter

... if output is also an argument or option. We still do want it to raise if it is only an output. We don't want it be set before validation so I think we still need line 36.

File: lib/flow/state/output.rb (hunk 1 above)
Comment (tbconroy) [reply lib/flow/state/output.rb:43]: Though I guess I could do [*_options,  *_arguments.keys].include?(method_name)

**Comment by tbconroy** [comment]
> @tbconroy we can't do that. Those modules are used by objects that aren't output objects, so that would create some pretty hairy interdependencies

@arettberg What should we do about the duplicate logic in SpiceRack that is also in `Flow::State::Output` and `Flow::State::Status`?

File: lib/flow/state/output.rb (hunk 1 above)
Comment (arettberg) [reply lib/flow/state/output.rb:43]: oh got it, yep that makes sense. I think the comment was a little confusing 😅

File: lib/flow/state/output.rb (hunk 2 above)
Comment (arettberg) [reply lib/flow/state/output.rb:41]: right right, sorry wasn't paying enough attention. can we maybe name the var `attr_name` instead for clarity?

**Comment by tbconroy** [comment]
Perhaps this could be closed in favor of https://github.com/Freshly/flow/pull/186?

**Comment by tbconroy** [comment]
We are going to fix this in Spicerack/Substance instead https://github.com/Freshly/spicerack/pull/471/files

**Pull request closed without merging**